In [1]:
print("Hello Analyst!")

Hello Analyst!


In [2]:
import pandas as pd


## Period 1 Daily Activity Data - Initial Inspection


- 457 rows and 14 columns
- 35 unique users
- Data covers 12 March 2016 to 12 April 2016 (32 calendar dates)
- No null values found
- No duplicate rows
- No duplicate Id/ActivityDate combinations
- Users contributed between 8 and 32 days of data
- Median user contributed 12 days of data
- ActivityDate was converted from string to datetime

In [3]:
df =pd.read_csv(r"C:\Users\ellio\Desktop\bellabeat_case_study\data\raw\fitbit_2016-03-12_to_2016-04-11\dailyActivity_merged.csv")



In [4]:
df.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,3/25/2016,11004,7.11,7.11,0.0,2.57,0.46,4.07,0.0,33,12,205,804,1819
1,1503960366,3/26/2016,17609,11.55,11.55,0.0,6.92,0.73,3.91,0.0,89,17,274,588,2154
2,1503960366,3/27/2016,12736,8.53,8.53,0.0,4.66,0.16,3.71,0.0,56,5,268,605,1944
3,1503960366,3/28/2016,13231,8.93,8.93,0.0,3.19,0.79,4.95,0.0,39,20,224,1080,1932
4,1503960366,3/29/2016,12041,7.85,7.85,0.0,2.16,1.09,4.61,0.0,28,28,243,763,1886


In [5]:
df.columns

Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='str')

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        457 non-null    int64  
 1   ActivityDate              457 non-null    str    
 2   TotalSteps                457 non-null    int64  
 3   TotalDistance             457 non-null    float64
 4   TrackerDistance           457 non-null    float64
 5   LoggedActivitiesDistance  457 non-null    float64
 6   VeryActiveDistance        457 non-null    float64
 7   ModeratelyActiveDistance  457 non-null    float64
 8   LightActiveDistance       457 non-null    float64
 9   SedentaryActiveDistance   457 non-null    float64
 10  VeryActiveMinutes         457 non-null    int64  
 11  FairlyActiveMinutes       457 non-null    int64  
 12  LightlyActiveMinutes      457 non-null    int64  
 13  SedentaryMinutes          457 non-null    int64  
 14  Calories             

In [7]:
df.describe()

,Id,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
count,4.570000e+02,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000,457.000000
mean,4.628595e+09,6546.562363,4.663523,4.609847,0.179427,1.180897,0.478643,2.890197,0.001904,16.623632,13.070022,170.070022,995.282276,2189.452954
std,2.293781e+09,5398.493064,4.082072,4.068540,0.849232,2.487159,0.830995,2.237523,0.008487,28.919704,36.208635,122.205372,337.021404,815.484523
min,1.503960e+09,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,32.000000,0.000000
25%,2.347168e+09,1988.000000,1.410000,1.280000,0.000000,0.000000,0.000000,0.870000,0.000000,0.000000,0.000000,64.000000,728.000000,1776.000000
50%,4.057193e+09,5986.000000,4.090000,4.090000,0.000000,0.000000,0.020000,2.930000,0.000000,0.000000,1.000000,181.000000,1057.000000,2062.000000
75%,6.391747e+09,10198.000000,7.160000,7.110000,0.000000,1.310000,0.670000,4.460000,0.000000,25.000000,16.000000,257.000000,1285.000000,2667.000000
max,8.877689e+09,28497.000000,27.530001,27.530001,6.727057,21.920000,6.400000,12.510000,0.100000,202.000000,660.000000,720.000000,1440.000000,4562.000000


In [8]:
df['Id'].nunique()

35

In [9]:
df["Id"].value_counts()

Id
4020332650    32
4057192912    32
1503960366    19
1624580081    19
2347167796    15
4445114986    15
4702921684    15
6962181067    14
1844505072    12
1927972279    12
2022484408    12
2026352035    12
2320127002    12
2873212765    12
3977333714    12
4319703577    12
4558609924    12
5553957443    12
7007744171    12
7086361926    12
8253242879    12
8378563200    12
8792009665    12
8877689391    12
5577150313    11
8053475328    11
1644430081    10
3372868164    10
6117666160    10
6290855005    10
6391747486     9
6775888955     9
2891001357     8
4388161847     8
8583815059     8
Name: count, dtype: int64

In [10]:
df["ActivityDate"] = pd.to_datetime(df["ActivityDate"])

In [11]:
df["ActivityDate"].dtype

dtype('<M8[us]')

In [12]:
df["ActivityDate"].min()

Timestamp('2016-03-12 00:00:00')

In [13]:
df["ActivityDate"].max()

Timestamp('2016-04-12 00:00:00')

In [14]:
df["ActivityDate"].max() - df["ActivityDate"].min()

Timedelta('31 days 00:00:00')

In [15]:
df["ActivityDate"].nunique()

32

In [16]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
452    False
453    False
454    False
455    False
456    False
Length: 457, dtype: bool

In [17]:
df.duplicated().sum()

np.int64(0)

In [18]:
df.duplicated(subset=["Id", "ActivityDate"]).sum()

np.int64(0)

In [19]:
user_days = df["Id"].value_counts()

In [20]:
user_days.describe()

count    35.000000
mean     13.057143
std       5.351981
min       8.000000
25%      10.500000
50%      12.000000
75%      12.000000
max      32.000000
Name: count, dtype: float64

In [21]:
sleep = pd.read_csv(r"C:\Users\ellio\Desktop\bellabeat_case_study\data\raw\fitbit_2016-03-12_to_2016-04-11\minuteSleep_merged.csv")

In [22]:
sleep.head(20)

,Id,date,value,logId
0,1503960366,3/13/2016 2:39:30 AM,1,11114919637
1,1503960366,3/13/2016 2:40:30 AM,1,11114919637
2,1503960366,3/13/2016 2:41:30 AM,1,11114919637
3,1503960366,3/13/2016 2:42:30 AM,1,11114919637
4,1503960366,3/13/2016 2:43:30 AM,1,11114919637
5,1503960366,3/13/2016 2:44:30 AM,1,11114919637
6,1503960366,3/13/2016 2:45:30 AM,2,11114919637
7,1503960366,3/13/2016 2:46:30 AM,2,11114919637
8,1503960366,3/13/2016 2:47:30 AM,1,11114919637
9,1503960366,3/13/2016 2:48:30 AM,1,11114919637


In [23]:
sleep.shape

(198559, 4)

In [24]:
sleep["value"].value_counts()

value
1    183526
2     13010
3      2023
Name: count, dtype: int64

## Period 1 Minute Sleep Data - Initial Inspection

- 198,559 rows and 4 columns
- One row represents one minute of sleep data for one user
- Sleep state values:
  - 1 = asleep
  - 2 = restless
  - 3 = awake
- 23 unique users have recorded sleep data, compared with 35 users in the daily activity dataset.
- 525 duplicate rows were identified and removed.
- No null values were found.

In [25]:
sleep["Id"].nunique()

23

In [26]:
sleep["date"].dtype

<StringDtype(storage='python', na_value=nan)>

In [27]:
sleep["date"] = pd.to_datetime(sleep["date"])

In [28]:
sleep["date"].dtype

dtype('<M8[us]')

In [29]:
sleep["date"].min()

Timestamp('2016-03-11 21:19:30')

In [30]:
sleep["date"].max()

Timestamp('2016-04-12 08:35:00')

In [31]:
sleep.duplicated().sum()

np.int64(525)

In [32]:
sleep[sleep.duplicated(keep=False)].head(10)

,Id,date,value,logId
79892,4319703577,2016-04-05 22:50:00,3,11344563687
79893,4319703577,2016-04-05 22:51:00,3,11344563687
79894,4319703577,2016-04-05 22:52:00,2,11344563687
79895,4319703577,2016-04-05 22:53:00,2,11344563687
79896,4319703577,2016-04-05 22:54:00,2,11344563687
79897,4319703577,2016-04-05 22:55:00,1,11344563687
79898,4319703577,2016-04-05 22:56:00,1,11344563687
79899,4319703577,2016-04-05 22:57:00,2,11344563687
79900,4319703577,2016-04-05 22:58:00,2,11344563687
79901,4319703577,2016-04-05 22:59:00,1,11344563687


In [33]:
sleep[
    (sleep["Id"] == 4319703577) &
    (sleep["date"] == "2016-04-05 22:50:00")
]

,Id,date,value,logId
79892,4319703577,2016-04-05 22:50:00,3,11344563687
80158,4319703577,2016-04-05 22:50:00,3,11344563687


In [34]:
sleep=sleep.drop_duplicates()

In [35]:
sleep.duplicated().sum()

np.int64(0)

In [36]:
sleep.isnull().sum()

Id       0
date     0
value    0
logId    0
dtype: int64

## Period 2 - Daily Sleep Data: Initial Inspection

**Period:** 12 April 2016 to 12 May 2016

- 413 rows and 5 columns initially
- 24 unique users
- 31 unique dates
- `SleepDay` converted from string to datetime
- No null values found
- 3 duplicate rows identified and removed
- 410 rows remained after duplicate removal
- Extreme low sleep-duration records were investigated and retained because there was insufficient evidence that they were erroneous

In [37]:
sleep_daily= pd.read_csv(r"C:\Users\ellio\Desktop\bellabeat_case_study\data\raw\fitbit_2016-04-12_to_2016-05-12\sleepDay_merged.csv")

In [38]:
sleep_daily.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,4/12/2016 12:00:00 AM,1,327,346
1,1503960366,4/13/2016 12:00:00 AM,2,384,407
2,1503960366,4/15/2016 12:00:00 AM,1,412,442
3,1503960366,4/16/2016 12:00:00 AM,2,340,367
4,1503960366,4/17/2016 12:00:00 AM,1,700,712


In [39]:
sleep_daily.shape

(413, 5)

In [40]:
sleep_daily["SleepDay"].dtype

<StringDtype(storage='python', na_value=nan)>

In [41]:
sleep_daily["SleepDay"] = pd.to_datetime(sleep_daily["SleepDay"])

C:\Users\ellio\AppData\Local\Temp\ipykernel_11640\3481005203.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  sleep_daily["SleepDay"] = pd.to_datetime(sleep_daily["SleepDay"])


In [42]:
sleep_daily["SleepDay"] = pd.to_datetime(
    sleep_daily["SleepDay"],
    format="%m/%d/%Y %I:%M:%S %p"
)

In [43]:
sleep_daily["SleepDay"].dtype

dtype('<M8[us]')

In [44]:
sleep_daily.info()

<class 'pandas.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Id                  413 non-null    int64         
 1   SleepDay            413 non-null    datetime64[us]
 2   TotalSleepRecords   413 non-null    int64         
 3   TotalMinutesAsleep  413 non-null    int64         
 4   TotalTimeInBed      413 non-null    int64         
dtypes: datetime64[us](1), int64(4)
memory usage: 16.3 KB


In [45]:
sleep_daily.duplicated().sum()

np.int64(3)

In [46]:
sleep_daily.isnull().sum()

Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

In [47]:
sleep_daily[sleep_daily.duplicated(keep=False)]

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
160,4388161847,2016-05-05,1,471,495
161,4388161847,2016-05-05,1,471,495
222,4702921684,2016-05-07,1,520,543
223,4702921684,2016-05-07,1,520,543
379,8378563200,2016-04-25,1,388,402
380,8378563200,2016-04-25,1,388,402


In [48]:
sleep_daily = sleep_daily.drop_duplicates()

In [49]:
sleep_daily.duplicated().sum()

np.int64(0)

In [50]:
sleep_daily.shape

(410, 5)

In [51]:
sleep_daily.nunique()

Id                     24
SleepDay               31
TotalSleepRecords       3
TotalMinutesAsleep    256
TotalTimeInBed        242
dtype: int64

In [52]:
sleep_daily["SleepDay"].max() - sleep_daily["SleepDay"].min()

Timedelta('30 days 00:00:00')

In [53]:
sleep_daily.describe()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
count,4.100000e+02,410,410.000000,410.000000,410.000000
mean,4.994963e+09,2016-04-26 11:38:55.609756,1.119512,419.173171,458.482927
min,1.503960e+09,2016-04-12 00:00:00,1.000000,58.000000,61.000000
25%,3.977334e+09,2016-04-19 00:00:00,1.000000,361.000000,403.750000
50%,4.702922e+09,2016-04-27 00:00:00,1.000000,432.500000,463.000000
75%,6.962181e+09,2016-05-04 00:00:00,1.000000,490.000000,526.000000
max,8.792010e+09,2016-05-12 00:00:00,3.000000,796.000000,961.000000
std,2.060863e+09,NaN,0.346636,118.635918,127.455140


In [54]:
mask = sleep_daily["TotalMinutesAsleep"] ==58
sleep_daily[mask]

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
338,7007744171,2016-05-01,1,58,61


In [56]:
sleep_daily.sort_values("TotalMinutesAsleep").head(10)

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
338,7007744171,2016-05-01,1,58,61
123,4319703577,2016-04-21,1,59,65
65,2320127002,2016-04-23,1,61,69
164,4388161847,2016-05-09,1,62,65
282,5577150313,2016-05-05,1,74,78
365,8053475328,2016-05-07,1,74,75
110,4020332650,2016-04-16,1,77,77
337,7007744171,2016-04-16,1,79,82
150,4388161847,2016-04-22,1,82,85
172,4445114986,2016-04-17,1,98,107


## Period 2 - Daily Activity Data: Initial Inspection

**Period:** 12 April 2016 to 12 May 2016

- 940 rows and 15 columns
- 33 unique users
- 31 calendar dates
- `ActivityDate` converted from string to datetime
- No null values found
- No duplicate rows found
- Users contributed between 4 and 31 days of activity data
- Median user contributed 31 days
- Mean user contributed approximately 28.5 days
- Most users had very complete activity data, although a small number had substantially fewer recorded days

### Data Quality Observations

- 79 records contained exactly 1,440 sedentary minutes (24 hours).
- 72 of these records also contained 0 total steps.
- 7 records contained 1,440 sedentary minutes despite recording between roughly 3,000 and 12,000 steps.
- These 7 records also showed 0 very active, fairly active, and lightly active minutes despite having substantial step counts and distances.
- This suggests that `SedentaryMinutes` and the activity-intensity minute fields may contain incomplete or inconsistent tracking for some records.
- These records will not automatically be removed because other fields, such as `TotalSteps`, distance, and calories, may still contain useful information.
- Activity-minute and sedentary-minute fields should therefore be interpreted cautiously in later analysis.
- 4 records contained 0 calories, which is physiologically implausible and likely reflects incomplete or missing tracking. These records should be treated cautiously in any calorie-based analysis.

In [88]:
activity_daily= pd.read_csv(r"C:\Users\ellio\Desktop\bellabeat_case_study\data\raw\fitbit_2016-04-12_to_2016-05-12\dailyActivity_merged.csv")

In [89]:
activity_daily.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,4/12/2016,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,4/13/2016,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,4/14/2016,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,4/15/2016,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,4/16/2016,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [90]:
activity_daily.info()

<class 'pandas.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDate              940 non-null    str    
 2   TotalSteps                940 non-null    int64  
 3   TotalDistance             940 non-null    float64
 4   TrackerDistance           940 non-null    float64
 5   LoggedActivitiesDistance  940 non-null    float64
 6   VeryActiveDistance        940 non-null    float64
 7   ModeratelyActiveDistance  940 non-null    float64
 8   LightActiveDistance       940 non-null    float64
 9   SedentaryActiveDistance   940 non-null    float64
 10  VeryActiveMinutes         940 non-null    int64  
 11  FairlyActiveMinutes       940 non-null    int64  
 12  LightlyActiveMinutes      940 non-null    int64  
 13  SedentaryMinutes          940 non-null    int64  
 14  Calories             

In [91]:
activity_daily["ActivityDate"] = pd.to_datetime(activity_daily["ActivityDate"])

In [92]:
activity_daily["ActivityDate"].dtype

dtype('<M8[us]')

In [93]:
activity_daily.duplicated().sum()

np.int64(0)

##Find the date range

In [94]:
activity_daily["ActivityDate"].max() - activity_daily["ActivityDate"].min()

Timedelta('30 days 00:00:00')

In [95]:
print(activity_daily["ActivityDate"].max())
print(activity_daily["ActivityDate"].min())
      

2016-05-12 00:00:00
2016-04-12 00:00:00


In [96]:
activity_daily["Id"].nunique()

33

In [97]:
activity_daily["Id"].value_counts().describe()

count    33.000000
mean     28.484848
std       5.657524
min       4.000000
25%      29.000000
50%      31.000000
75%      31.000000
max      31.000000
Name: count, dtype: float64

In [98]:
activity_daily.describe()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
count,9.400000e+02,940,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000,940.000000
mean,4.855407e+09,2016-04-26 06:53:37.021276,7637.910638,5.489702,5.475351,0.108171,1.502681,0.567543,3.340819,0.001606,21.164894,13.564894,192.812766,991.210638,2303.609574
min,1.503960e+09,2016-04-12 00:00:00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2.320127e+09,2016-04-19 00:00:00,3789.750000,2.620000,2.620000,0.000000,0.000000,0.000000,1.945000,0.000000,0.000000,0.000000,127.000000,729.750000,1828.500000
50%,4.445115e+09,2016-04-26 00:00:00,7405.500000,5.245000,5.245000,0.000000,0.210000,0.240000,3.365000,0.000000,4.000000,6.000000,199.000000,1057.500000,2134.000000
75%,6.962181e+09,2016-05-04 00:00:00,10727.000000,7.712500,7.710000,0.000000,2.052500,0.800000,4.782500,0.000000,32.000000,19.000000,264.000000,1229.500000,2793.250000
max,8.877689e+09,2016-05-12 00:00:00,36019.000000,28.030001,28.030001,4.942142,21.920000,6.480000,10.710000,0.110000,210.000000,143.000000,518.000000,1440.000000,4900.000000
std,2.424805e+09,NaN,5087.150742,3.924606,3.907276,0.619897,2.658941,0.883580,2.040655,0.007346,32.844803,19.987404,109.174700,301.267437,718.166862


In [99]:
activity_daily.sort_values("TotalSteps").head(10)

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
908,8792009665,2016-05-10,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,48,57
654,6775888955,2016-04-12,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1841
642,6290855005,2016-04-29,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2060
30,1503960366,2016-05-12,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,0
893,8792009665,2016-04-25,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1688
742,7086361926,2016-04-17,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1629
347,4020332650,2016-04-13,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1981
610,6117666160,2016-04-25,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1497
104,1844505072,2016-04-24,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1347
663,6775888955,2016-04-21,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1841


In [100]:
mask1 = activity_daily["SedentaryMinutes"] == 1440

In [101]:
mask1.sum()

np.int64(79)

In [102]:
activity_daily[mask1]

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
30,1503960366,2016-05-12,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,0
104,1844505072,2016-04-24,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1347
105,1844505072,2016-04-25,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1347
106,1844505072,2016-04-26,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1347
112,1844505072,2016-05-02,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1348
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
903,8792009665,2016-05-05,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1688
904,8792009665,2016-05-06,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1688
905,8792009665,2016-05-07,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1688
906,8792009665,2016-05-08,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,1688


In [103]:
mask2 = (activity_daily["TotalSteps"] == 0) & (activity_daily["SedentaryMinutes"] == 1440)

In [104]:
mask2.sum()

np.int64(72)

In [105]:
mask3 = (activity_daily["TotalSteps"] > 0) & (activity_daily["SedentaryMinutes"] == 1440)

In [106]:
activity_daily[mask3]

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
381,4319703577,2016-04-12,7753,5.20,5.20,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2115
412,4388161847,2016-04-12,10122,7.78,7.78,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2955
853,8583815059,2016-04-16,5319,4.15,4.15,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2693
854,8583815059,2016-04-17,3008,2.35,2.35,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2439
869,8583815059,2016-05-02,8469,6.61,6.61,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2894
870,8583815059,2016-05-03,12015,9.37,9.37,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,3212
871,8583815059,2016-05-04,3588,2.80,2.80,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,2516


In [107]:
activity_daily["Calories"].describe()

count     940.000000
mean     2303.609574
std       718.166862
min         0.000000
25%      1828.500000
50%      2134.000000
75%      2793.250000
max      4900.000000
Name: Calories, dtype: float64

In [108]:
(activity_daily["Calories"] ==0).sum()

np.int64(4)

### Analysis Scope

Period 2 (12 April 2016 to 12 May 2016) was selected as the primary dataset for analysis.

Period 2 provides more complete participant coverage and contains both daily activity and daily sleep data. Period 1 was inspected during the data preparation stage but was not included in the main analysis due to its lower data completeness.

Focusing on Period 2 reduces unnecessary complexity while still providing a full month of Fitbit data from which meaningful user behaviour trends can be explored.

In [110]:
activity_daily["Id"].nunique()

33

In [111]:
sleep_daily["Id"].nunique()

24

In [ ]:
activity_daily.loc[
    activity_daily["Id"].isin(sleep_daily["Id"]),"Id"
].nunique()

24

In [116]:
activity_daily["ActivityDate"].head()

0   2016-04-12
1   2016-04-13
2   2016-04-14
3   2016-04-15
4   2016-04-16
Name: ActivityDate, dtype: datetime64[us]

In [117]:
sleep_daily["SleepDay"].head()

0   2016-04-12
1   2016-04-13
2   2016-04-15
3   2016-04-16
4   2016-04-17
Name: SleepDay, dtype: datetime64[us]

# Analyse

### Analysis Plan

The analysis will focus on Period 2 data (12 April–12 May 2016).

Main areas to explore:
- Daily activity patterns, primarily using `TotalSteps`
- Daily sleep patterns using `TotalMinutesAsleep`
- Differences in activity and sleep across days of the week
- A simple exploration of the relationship between sleep duration and activity, if the data can be aligned appropriately

The analysis will prioritise clear, high-level trends rather than granular activity-intensity measures due to identified data-quality limitations.